runtime/sam/op/robot/robot.go

```diff
@@ -0,0 +1,193 @@
+		}
+		o.batch = nil
+		src := o.src
+		o.src = nil
+		if closer, ok := src.(io.Closer); ok {
+			if err := closer.Close(); err != nil {
+				return nil, err
+			}
+		}
+		b, err := o.parent.Pull(true)
+		if b != nil {
+			b.Unref()
+		}
+		return nil, err
+	}
+	for {
+		puller, err := o.getPuller()
+		if puller == nil || err != nil {
+			o.src = nil
+			return nil, err
+		}
+		b, err := puller.Pull(false)
+		if b != nil || err != nil {
+			return b, err
+		}
+		if b == nil {
+			o.src = nil
+		}
+	}
+}
+
+func (o *Op) getPuller() (zbuf.Puller, error) {
+	puller := o.src
+	if puller != nil {
+		return puller, nil
+	}
+	if len(o.targets) > 0 {
+		src, err := o.openNext()
+		o.src = src
+		return src, err
+	}
+	src, err := o.nextPuller()
+	o.src = src
+	return src, err
+}
+
+func (o *Op) nextPuller() (zbuf.Puller, error) {
+	b := o.batch
+	if b != nil && o.off >= len(b.Values()) {
+		b.Unref()
+		o.off = 0
+		o.batch = nil
+		b = nil
+	}
+	if b == nil {
+		var err error
+		b, err = o.nextBatch()
+		if err != nil {
+			return nil, err
+		}
+		o.batch = b
+		o.off = 0
+		if b == nil {
+			return nil, nil
+		}
+	}
+	val := b.Values()[o.off]
+	o.off++
+	return o.openFromVal(b, val)
+}
+
+func (o *Op) openFromVal(b zbuf.Batch, val super.Value) (zbuf.Puller, error) {
+	target := o.expr.Eval(b, val)
+	typ := super.TypeUnder(target.Type())
+	if typ == super.TypeString {
+		return o.open(target.AsString())
+	}
+	vals, err := target.Elements()
+	if err != nil || len(vals) == 0 {
+		return o.errOnVal(target), nil
+	}
+	typ = super.TypeUnder(vals[0].Type())
+	if typ != super.TypeString {
+		return o.errOnVal(target), nil
+	}
+	o.targets = vals
+	return o.openNext()
+}
+
+func (o *Op) openNext() (zbuf.Puller, error) {
+	if len(o.targets) == 0 {
+		return nil, nil
+	}
+	val := o.targets[0]
+	o.targets = o.targets[1:]
+	return o.open(val.AsString())
+}
+
+func (o *Op) errOnVal(val super.Value) zbuf.Puller {
+	errVal := o.rctx.Zctx.WrapError("from encountered non-string input", val)
+	return zbuf.NewPuller(zbuf.NewArray([]super.Value{errVal}))
+}
+
+func (o *Op) nextBatch() (zbuf.Batch, error) {
+again:
+	b, err := o.parent.Pull(false)
+	if err != nil {
+		return nil, err
+	}
+	if b == nil {
+		o.batch = nil
+		o.off = 0
+		return nil, nil
+	}
+	if len(b.Values()) == 0 {
+		b.Unref()
+		goto again
+	}
+	o.batch = b
+	o.off = 0
+	return b, nil
+}
+
+func (o *Op) open(path string) (zbuf.Puller, error) {
+	u, err := storage.ParseURI(path)
+	if err == nil && false {
+		//XXX get from AST args, or we can also get this stuff from the
+		// robot expr, e.g., allowning the querying to create a record
+		// to hold these args...
+		var method string
+		var body io.Reader
+		var headers http.Header
+		f, err := o.env.OpenHTTP(o.rctx.Context, o.rctx.Zctx, u.String(), o.format, method, headers, body, nil)
+		if err != nil {
+			return nil, err
+		}
+		return f, err
+	}
+	if o.env.IsLake() {
+		return nil, fmt.Errorf("%s: cannot open in a data lake environment", path)
+	}
```

Comment:
This check could happen in semantic.analyzer.semFromExpr.

Reply:
It's gonna go away when we add support for lake's in robotscan.